src/ray/gcs/redis_context.cc

```diff
@@ -563,10 +462,15 @@ std::unique_ptr<CallbackReply> RedisContext::RunArgvSync(
 
 void RedisContext::RunArgvAsync(std::vector<std::string> args,
                                 RedisCallback redis_callback) {
-  RAY_CHECK(redis_async_context_);
+  std::shared_ptr<RedisAsyncContext> redis_async_context;
+  {
+    absl::MutexLock l(&mu_);
+    RAY_CHECK(redis_async_context_);
+    redis_async_context = redis_async_context_;
+  }
```

Comment:
why lock and copy?

Reply:
I guess it is okay not to do it this way as long as the `request_context->Run()` is not in the critical section... I can change it to direct copy? Doesn't matter much to me

Reply:
Basically the model should be: Each RedisRequestContext gets a refcounted copy of the async_context. When there is a MOVED, the async_context should be changed, as well as the context kept in the parent RedisContext. Lifetime of the parent context is tied to the lifetime of the RedisClient. 

Already-created RedisRequestContexts will still have the old async_context, so they will each do the extra work of redirecting. However, after the first replacement, any newly created RedisRequestContexts will have the updated context to the correct node.